### ISSUE TITLE
KeyError session is disconnected

### ISSUE THREAD

**Issue by plockaby** [miguelgrinberg/python-engineio#83]
Hi Miguel, sorry to bother you again. I've got another one that I'm not quite sure what to do with because I can't actually catch the exception:

```
[2018-11-01 16:46:51,822] DEBUG    - POST /events/stream/
[2018-11-01 16:46:51,823] ERROR    - Error handling request /events/stream/?EIO=3&transport=polling&t=MRHh25Q&sid=304b7ca5ed064bc49d98eb4036263b9b
Traceback (most recent call last):
  File "/netops/venv/events-portal-web/lib/python3.6/site-packages/gunicorn/workers/base_async.py", line 56, in handle
    self.handle_request(listener_name, req, client, addr)
  File "/netops/venv/events-portal-web/lib/python3.6/site-packages/gunicorn/workers/base_async.py", line 107, in handle_request
    respiter = self.wsgi(environ, resp.start_response)
  File "/netops/venv/events-portal-web/lib/python3.6/site-packages/flask/app.py", line 2309, in __call__
    return self.wsgi_app(environ, start_response)
  File "/netops/venv/events-portal-web/lib/python3.6/site-packages/flask_socketio/__init__.py", line 43, in __call__
    start_response)
  File "/netops/venv/events-portal-web/lib/python3.6/site-packages/engineio/middleware.py", line 47, in __call__
    return self.engineio_app.handle_request(environ, start_response)
  File "/netops/venv/events-portal-web/lib/python3.6/site-packages/socketio/server.py", line 360, in handle_request
    return self.eio.handle_request(environ, start_response)
  File "/netops/venv/events-portal-web/lib/python3.6/site-packages/engineio/server.py", line 298, in handle_request
    socket = self._get_socket(sid)
  File "/netops/venv/events-portal-web/lib/python3.6/site-packages/engineio/server.py", line 439, in _get_socket
    raise KeyError('Session is disconnected')
KeyError: 'Session is disconnected'
[2018-11-01 16:46:52,503] INFO     - emitting event "update" to all [/]
```

Thoughts about how I might handle this?

I updated this to add a bit more context around the error. I have a background thread in Flask that every few seconds sends out a message to everyone with new data. It appears that a client may have been trying to connect in the middle of me sending an update. But it's not clear.

Also in reference to the previous issue that I brought up, that seems to have gone away but may have been related to this background thread and my application's connection to Cassandra. Again, it looks like that has gone away with your changes and with some changes I made to my Cassandra connection handler. Thanks again for your help.

**Comment by miguelgrinberg** [comment]
The `session is disconnected` error is not something you can catch yourself. The problem occurs when a client does not timely send ping packets to the server, as indicated by the Socket.IO protocol. A client should send a ping packet every 25 seconds (in the default configuration). If the server does not receive a ping in 60 seconds it declares the client gone and closes that connection. If after 60 seconds, the client comes back and tries to send a ping you'll get the "session is disconnected" error, because the server already gave up on that client and fully released it.

So basically you need to research why your clients do not send proper pings, or alternatively, if the server is maybe blocking for long periods of time, which would prevent it from handling pings in a timely manner.

**Comment by miguelgrinberg** [comment]
Closing due to inactivity. Please reopen if this is still a problem.

**Comment by plockaby** [comment]
Oh sorry. I forgot about this. The actual problem was caused by a bad import that was breaking the monkey patching.
